src/lib/pipeline/LoadStoreQueue.cc

```diff
@@ -422,55 +422,57 @@ void LoadStoreQueue::tick() {
-            if (dataTransfered[isStore] > bandwidth) {
-              // No more requests can be scheduled this cycle
-              exceededLimits[isStore] = true;
-              itInsn = itReq->second.end();
-              break;
-            }
-
-            // Request a read from the memory interface if the requestQueue_
-            // entry represents a read
-            if (!isStore) {
-              memory_.requestRead(req, itInsn->insn->getSequenceId());
-            }
+        // Schedule requests from the queue of addresses in
+        // request[Load|Store]Queue_ entry
+        auto& addressQueue = itInsn->reqAddresses;
+        while (addressQueue.size()) {
+          const simeng::MemoryAccessTarget req =
+              addressQueue.front();  // Speculatively increment count of this
+                                     // request type
+          reqCounts[isStore]++;
+
+          // Ensure the limit on the number of permitted operations is adhered
+          // to
+          if (reqCounts[isStore] + reqCounts[!isStore] > totalLimit_) {
+            // No more requests can be scheduled this cycle
+            exceededLimits = {true, true};
+            itInsn = itReq->second.end();
+            break;
+          } else if (reqCounts[isStore] > reqLimits_[isStore]) {
+            // No more requests of this type can be scheduled this cycle
+            exceededLimits[isStore] = true;
+            // Remove speculative increment to ensure it doesn't count for
+            // comparisons aginast the totalLimit_
+            reqCounts[isStore]--;
+            itInsn = itReq->second.end();
+            break;
+          }
 
-            // Remove processed address from queue
-            addressQueue.pop();
+          // Ensure the limit on the data transfered per cycle is adhered to
+          assert(req.size <= bandwidth &&
+                 "Individual memory request from LoadStoreQueue exceeds L1 "
+                 "bandwidth set and thus will never be submitted");
+          dataTransfered[isStore] += req.size;
+          if (dataTransfered[isStore] > bandwidth) {
+            // No more requests can be scheduled this cycle
+            exceededLimits[isStore] = true;
+            itInsn = itReq->second.end();
+            break;
           }
-          // Remove entry from vector iff all of its requests have been
-          // scheduled
-          if (addressQueue.size() == 0) {
-            itInsn = itReq->second.erase(itInsn);
+
+          // Request a read from the memory interface if the requestQueue_
+          // entry represents a read
+          if (!isStore) {
+            memory_.requestRead(req, itInsn->insn->getSequenceId());
           }
+
+          // Remove processed address from queue
+          addressQueue.pop();
+        }
+        // Remove entry from vector iff all of its requests have been
+        // scheduled
+        if (addressQueue.size() == 0) {
+          itInsn = itReq->second.erase(itInsn);
```

Comment:
I'm not 100% sure on why this needed changing? Are you able to briefly explain why? Thanks

Reply:
This change was done so that the limits were applied to individual address accesses rather than individual uops. We could have a case whereby a uop with 8 accesses (8 separate memory flows) would be considered as a single access from the LSQ restrictions pov.

Reply:
I see, so for instructions like an FP32 Gather, previously it would treat this as 1 access rather than the 8 individual accesses that it should do (assuming 8 elements per FP32 vector)?

And just to clarify - a normal SVE vector load (assuming `ptrue p0`) will still be done as a single access / flow?

Reply:
Yes exactly. The restrictions were being placed on the entries in `requestLoadQueue_` or `requestStoreQueue_` rather than the addresses in each entry. Given our updates to the SVE contiguous accesses, they only have 1 address in each of those request queue entries so yes, 1 flow